src/backend/renderer/utils.rs

```diff
@@ -19,31 +18,70 @@ use wayland_server::protocol::{wl_buffer::WlBuffer, wl_surface::WlSurface};
 
 #[derive(Default)]
 pub(crate) struct SurfaceState {
+    pub(crate) commit_count: usize,
     pub(crate) buffer_dimensions: Option<Size<i32, Buffer>>,
     pub(crate) buffer_scale: i32,
     pub(crate) buffer_transform: Transform,
     pub(crate) buffer: Option<WlBuffer>,
+    pub(crate) damage: VecDeque<Vec<Rectangle<i32, Buffer>>>,
+    pub(crate) renderer_seen: HashMap<(TypeId, usize), usize>,
     pub(crate) textures: HashMap<(TypeId, usize), Box<dyn std::any::Any>>,
     #[cfg(feature = "desktop")]
-    pub(crate) damage_seen: HashSet<crate::desktop::space::SpaceOutputHash>,
+    pub(crate) space_seen: HashMap<crate::desktop::space::SpaceOutputHash, usize>,
 }
 
+const MAX_DAMAGE: usize = 4;
+
 impl SurfaceState {
     pub fn update_buffer(&mut self, attrs: &mut SurfaceAttributes) {
         match attrs.buffer.take() {
             Some(BufferAssignment::NewBuffer { buffer, .. }) => {
                 // new contents
                 self.buffer_dimensions = buffer_dimensions(&buffer);
+
+                #[cfg(feature = "desktop")]
+                if self.buffer_scale != attrs.buffer_scale
+                    || self.buffer_transform != attrs.buffer_transform.into()
+                {
+                    self.reset_space_damage();
+                }
                 self.buffer_scale = attrs.buffer_scale;
                 self.buffer_transform = attrs.buffer_transform.into();
+
                 if let Some(old_buffer) = std::mem::replace(&mut self.buffer, Some(buffer)) {
                     if &old_buffer != self.buffer.as_ref().unwrap() {
                         old_buffer.release();
                     }
                 }
                 self.textures.clear();
-                #[cfg(feature = "desktop")]
-                self.damage_seen.clear();
+                let (new_commit_count, overflow) = self.commit_count.overflowing_add(1);
+                self.commit_count = new_commit_count;
+                if overflow {
+                    // we need to avoid bugs in `damage_since` (see below)
+                    self.damage.clear();
```

Comment:
In this code path, `attrs.damage` is not even read, is that expected?

Also, looking at `damage_since()`, it appears that function will always return and empty `Vec` if `self.damage` is empty, so it seems to me that we are effectively discarding all damage information on overflow, shouldn't there be at least an addition of a full damage over the surface or something like that?

Reply:
It will only return an empty vector, if the buffer dimensions are not set (and thus we have no buffer). Otherwise it will return full-damage, because damage.len() will be too small, resulting in `recent_enough` evaluating to false.

> In this code path, attrs.damage is not even read, is that expected?

But this is a good catch, this code path should probably clear attrs.damage.

Reply:
> Otherwise it will return full-damage, because damage.len() will be too small, resulting in recent_enough evaluating to false.

Oh, right, ok. Maybe let's add a comment on the computation of `recent_enough` explaining that? That in case of overflow, we'll actually have `self.commit_count < commit`